ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BeaconStateUtil.java

```diff
@@ -442,4 +556,285 @@ public static int clamp(int minval, int maxval, int x) {
+
+    List<Bytes48> signature =
+        Arrays.asList(
+            Bytes48.leftPad(proof_of_possession.get(0)),
+            Bytes48.leftPad(proof_of_possession.get(1)));
+    UnsignedLong domain = get_domain(state.getFork(), state.getSlot(), DOMAIN_DEPOSIT);
+    return bls_verify(
+        pubkey, TreeHashUtil.hash_tree_root(proof_of_possession_data.toBytes()), signature, domain);
+  }
+
+  /**
+   * @param fork
+   * @param slot
+   * @param domain_type
+   * @return
+   */
+  private static UnsignedLong get_domain(Fork fork, UnsignedLong slot, int domain_type) {
+    return get_fork_version(fork, slot)
+        .times(UnsignedLong.valueOf((long) Math.pow(2, 32)))
+        .plus(UnsignedLong.valueOf(domain_type));
+  }
+
+  /**
+   * @param fork
+   * @param epoch
+   * @return
+   */
+  public static UnsignedLong get_fork_version(Fork fork, UnsignedLong epoch) {
+    if (epoch.compareTo(fork.getEpoch()) < 0) return fork.getPrevious_version();
+    else return fork.getCurrent_version();
+  }
+
+  /**
+   * Returns the participant indices at for the 'attestation_data' and 'participation_bitfield'.
+   *
+   * @param state
+   * @param attestation_data
+   * @param participation_bitfield
+   * @return
+   */
+  public static ArrayList<Integer> get_attestation_participants(
+      BeaconState state, AttestationData attestation_data, byte[] participation_bitfield) {
+    // Find the relevant committee
+
+    ArrayList<CrosslinkCommittee> crosslink_committees =
+        BeaconStateUtil.get_crosslink_committees_at_slot(
+            state, toIntExact(attestation_data.getSlot()));
+
+    // TODO: assert attestation_data.shard in [shard for _, shard in crosslink_committees]
+
+    CrosslinkCommittee crosslink_committee = null;
+    for (CrosslinkCommittee curr_crosslink_committee : crosslink_committees) {
+      if (curr_crosslink_committee.getShard().equals(attestation_data.getShard())) {
+        crosslink_committee = curr_crosslink_committee;
+        break;
+      }
+    }
+    assert crosslink_committee != null;
+    assert participation_bitfield.length == ceil_div8(crosslink_committee.getCommitteeSize());
+
+    // Find the participating attesters in the committee
+    ArrayList<Integer> participants = new ArrayList<>();
+    for (int i = 0; i < crosslink_committee.getCommitteeSize(); i++) {
+      int participation_bit = (participation_bitfield[i / 8] >> (7 - (i % 8))) % 2;
+      if (participation_bit == 1) {
+        participants.add(crosslink_committee.getCommittee().get(i));
+      }
+    }
+    return participants;
+  }
+
+  public static Bytes32 generate_seed(BeaconState state, UnsignedLong epoch) {
+    return get_randao_mix(state, epoch.plus(UnsignedLong.valueOf(SEED_LOOKAHEAD)))
+        .and(get_active_index_root(state, epoch));
+  }
+
+  /**
+   * @param validators
+   * @param current_slot
+   * @return The minimum empty validator index.
+   */
+  private int min_empty_validator_index(
```

Comment:
I don't think this is in v0.1 either, so we might wanna just take it out.

Reply:
I agree but, I will leave that to someone else.